### PR TITLE
Don't use progressbars for pull if client is no-tty

### LINKED ISSUE #3694: docker pull without crazytown control code central
Take:

```
docker pull mattdm/fedora > test_it
```

then 

```
sam@ubuntu:~$ ls -al test_it 
-rw-rw-r-- 1 sam sam 32969 Jan 20 20:26 test_it
```

image

This format though wonderful from the console, is very hostile to scripting, it makes it completely impossible to track progress from a log file. 

Instead, offer a flag (which is also applicable to docker run since it does a pull) which allows for the following format:

```
Pulling repository mattdm/fedora
7c8cf65e1efa: Pulling image (f19) from mattdm/fedora
373dc830edcd:: Pulling image (f19) from mattdm/fedora
#######################################################
7c8cf65e1efa: Image downloaded (f19) from mattdm/fedora
#############################################################
373dc830edcd: Image downloaded (f19) from mattdm/fedora
```

One # for every say 256 kb downloaded

### PR DESCRIPTION
**Pull request by LK4D4**
Fix #3694 
It's only for design review, if it'll be ok, I'll do push as well.

### CODE REVIEW

**Comment by calavera** [comment]
I think this is a good idea. I'd like to know what @aaronlehmann and @tonistiigi think about it, I know they have some thoughts about the pull command.

**Comment by tonistiigi** [comment]
I'm ok with this atm but in the future the pull code should only handle progress messages with channels that transfer objects. So if this work is done this `isTerminal` switch would move to the api side.

**Comment by aaronlehmann** [comment]
I'm okay with this as well. As an enhancement, it would be nice to still output progress in a less awkward way if stdout is not a terminal. I'm working on an overhaul of the pull/push code that uses channels to communicate abstract progress information instead of writing JSON directly, which will make this a lot easier.

**Comment by LK4D4** [comment]
@aaronlehmann will that overhaul happen before 1.10?

**Comment by aaronlehmann** [comment]
Yeah, I plan to open the PR soon

**Comment by calavera** [comment]
@LK4D4, @aaronlehmann should we close this PR now that #18353 is open? Does that one address this issue?

**Comment by LK4D4** [comment]
@calavera no, it's not, I'm waiting for it to update this.

**Comment by LK4D4** [comment]
@calavera @aaronlehmann I updated PR.
However I see something weird without tty:
```
Using default tag: latest
latest: Pulling from library/busybox
d7e8ec85c5ab: Pulling fs layer
a3ed95caeb02: Pulling fs layer
a3ed95caeb02: Download complete
d7e8ec85c5ab: Download complete
d7e8ec85c5ab: Pull complete
d7e8ec85c5ab: Pull complete
a3ed95caeb02: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:e4f93f6ed15a0cdd342f5aae387886fba0ab98af0a102da6276eaf24d6e6ade0
Status: Downloaded newer image for busybox:latest
```
You can see that "Pull complete" appear twice for each layer.

**Comment by aaronlehmann** [comment]
Hmm. It's a side effect of my transfer manager PR. It does progress notification in an async way so that it doesn't block the actual transfers. This means some updates might be missed. When a transfer finishes, it replays the final update in case it was missed earlier.

Obviously, this was designed with the tty in mind, and it leads to weird output without a tty.

if you want, I can try to revisit this.

File: daemon/daemon.go [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ func (daemon *Daemon) PushImage(ref reference.Named, metaHeaders map[string][]st
 	ctx, cancelFunc := context.WithCancel(context.Background())
 
 	go func() {
-		writeDistributionProgress(cancelFunc, outStream, progressChan)
+		writeDistributionProgress(cancelFunc, outStream, progressChan, true)
```
Comment (aaronlehmann) [inline daemon/daemon.go]: Why not do this for push as well?

**Comment by aaronlehmann** [comment]
@LK4D4 Does https://github.com/aaronlehmann/docker/commit/417bb5caf19ae0d9e13a7379b10bbb845aa16a87 solve the problem? If so, feel free to include it in this PR.

**Comment by LK4D4** [comment]
@aaronlehmann Will try, thanks!

**Comment by LK4D4** [comment]
@aaronlehmann yup, it fixed it, but now I also see one `Verifying checksum` :)

**Comment by aaronlehmann** [comment]
I think that is expected.

**Comment by LK4D4** [comment]
@aaronlehmann Cool. Are you ok with feature overall?

**Comment by LK4D4** [comment]
However, I think this PR is unnecessary, it works in master.

**Pull request closed without merging**